interfaces/prompting/requestprompts/requestprompts.go

```diff
@@ -0,0 +1,351 @@
+//
+// The caller must ensure that the prompt DB mutex is held.
+func (pdb *PromptDB) nextID() string {
+	pdb.maxID += 1
+	padded := pdb.paddedMaxIDString()
+	osutil.AtomicWriteFile(pdb.maxIDPath, []byte(padded), 0600, 0)
+	return padded
+}
+
+// paddedMaxIDString returns a 16-character string corresponding to the current
+// maxID. The ID string is the max ID in hexadecimal, padded by leading zeroes.
+func (pdb *PromptDB) paddedMaxIDString() string {
+	maxIDStr := strconv.FormatUint(pdb.maxID, 16)
+	// pad with leading zeros
+	padded := "0000000000000000"[:16-len(maxIDStr)] + maxIDStr
+	return padded
+}
+
+// AddOrMerge checks if the given prompt contents are identical to an existing
+// prompt and, if so, merges with it by adding the given listenerReq to it.
+// Otherwise, adds a new prompt with the given contents to the prompt DB.
+//
+// If the prompt was merged with an identical existing prompt, returns the
+// existing prompt and true, indicating it was merged. If a new prompt was
+// added, returns the new prompt and false, indicating the prompt was not
+// merged.
+//
+// The caller must ensure that the given permissions are in the order in which
+// they appear in the available permissions list for the given interface.
+func (pdb *PromptDB) AddOrMerge(user uint32, snap string, iface string, path string, permissions []string, listenerReq *listener.Request) (*Prompt, bool) {
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	userEntry, exists := pdb.perUser[user]
+	if !exists {
+		pdb.perUser[user] = &userPromptDB{
+			ByID: make(map[string]*Prompt),
+		}
+		userEntry = pdb.perUser[user]
+	}
+
+	availablePermissions, _ := prompting.AvailablePermissions(iface)
+	// Error should be impossible, since caller has already validated that iface
+	// is valid, and tests check that all valid interfaces have valid available
+	// permissions returned by AvailablePermissions.
+
+	constraints := &promptConstraints{
+		Path:                 path,
+		Permissions:          permissions,
+		AvailablePermissions: availablePermissions,
+	}
+
+	// Search for an identical existing prompt, merge if found
+	for _, prompt := range userEntry.ByID {
+		if prompt.Snap == snap && prompt.Interface == iface && prompt.Constraints.equals(constraints) {
+			prompt.listenerReqs = append(prompt.listenerReqs, listenerReq)
+			return prompt, true
+		}
+	}
+
+	id := pdb.nextID()
+	timestamp := time.Now()
+	prompt := &Prompt{
+		ID:           id,
+		Timestamp:    timestamp,
+		Snap:         snap,
+		Interface:    iface,
+		Constraints:  constraints,
+		listenerReqs: []*listener.Request{listenerReq},
+	}
+	userEntry.ByID[id] = prompt
+	pdb.notifyPrompt(user, id)
+	return prompt, false
+}
+
+// Prompts returns a slice of all outstanding prompts.
+func (pdb *PromptDB) Prompts(user uint32) []*Prompt {
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	userEntry, exists := pdb.perUser[user]
+	if !exists {
+		return make([]*Prompt, 0)
```

Comment:
Why not return nil? Otherwise, this is the same without calling `make`
```suggestion
		return []*Prompt{}
```

Reply:
We don't want nil because we want json to unmarshal into an empty list, rather than into null. I'll use your suggestion to avoid explicitly calling `make` though, thanks!